File: Problem_2 Snake_game.py
# // Time Complexity : O(HW) where W and H are width and height respectively
# // Space Complexity : O(1)
# // Did this code successfully run on Leetcode : Yes
# // Any problem you faced while coding this : No
#
#
# // Your code here along with comments explaining your approach

class SnakeGame:

    def __init__(self, width: int, height: int, food: List[List[int]]):
        """
        Initialize your data structure here.
        @param width - screen width
        @param height - screen height
        @param food - A list of food positions
        E.g food = [[1,1], [1,0]] means the first food is positioned at [1,1], the second is at [1,0].
        """
        self.snake = deque([[0, 0]])
        self.width = width
        self.height = height
        self.food = food
        self.bitecheck = set()  # Hashset

        self.bitecheck.add((0, 0))
        self.index = 0
        self.dirs = {"D": [1, 0],
                     "U": [-1, 0],
                     "R": [0, 1],
                     "L": [0, -1]}

    def move(self, direction: str) -> int:
        """
        Moves the snake.
        @param direction - 'U' = Up, 'L' = Left, 'R' = Right, 'D' = Down
        @return The game's score after the move. Return -1 if game over.
        Game over when snake crosses the screen boundary or bites its body.
        """
        # Check for new head
        curr_head = self.snake[-1]
        update_head = [0, 0]
        update_head[0] = curr_head[0] + self.dirs[direction][0]
        update_head[1] = curr_head[1] + self.dirs[direction][1]

        row, col = update_head

        # if the snake hits the boundries or it touches itself, return -1
        if row < 0 or row >= self.height or col < 0 or col >= self.width or (
                (row, col) in self.bitecheck and [row, col] != self.snake[0]):
            return -1

        elif self.index < len(self.food) and self.food[self.index] == update_head:
            self.index += 1
        else:
            curr_tail = self.snake.popleft()
            r, c = curr_tail
            self.bitecheck.remove((r, c))

        self.bitecheck.add((row, col))
        self.snake.append(update_head)
        return len(self.snake) - 1

# Your SnakeGame object will be instantiated and called as such:
# obj = SnakeGame(width, height, food)
# param_1 = obj.move(direction)

File: Problem2_SnakeGame.java
//Time Complexity :o(N)
//Space Complexity : o(N)
//Did this code successfully run on Leetcode : Yes
//Any problem you faced while coding this : No
class SnakeGame {

    /** Initialize your data structure here.
        @param width - screen width
        @param height - screen height
        @param food - A list of food positions
        E.g food = [[1,1], [1,0]] means the first food is positioned at [1,1], the second is at [1,0]. */

    LinkedList<int[]> snake;
    LinkedList<int[]> foodList;
    int width;
    int height;
    int[] snakeHead;
    public SnakeGame(int width, int height, int[][] food) {
        this.width = width;
        this.height = height;
        snake = new LinkedList<>();
        foodList = new LinkedList<>(Arrays.asList(food));
        snakeHead = new int[]{0,0};
        snake.add(snakeHead);
    }

    /** Moves the snake.
        @param direction - 'U' = Up, 'L' = Left, 'R' = Right, 'D' = Down
        @return The game's score after the move. Return -1 if game over.
        Game over when snake crosses the screen boundary or bites its body. */
    public int move(String direction) {
        if(direction.equals("U")) {
            snakeHead[0]--;
        }
        if(direction.equals("D")) {
            snakeHead[0]++;
        }
        if(direction.equals("L")) {
            snakeHead[1]--;
        }
        if(direction.equals("R")) {
            snakeHead[1]++;
        }

        //Hits wall
        if(snakeHead[0] == height || snakeHead[0] < 0 || snakeHead[1] == width || snakeHead[1] < 0) {
            return -1;
        }

        //Hits itself
        for(int i = 1; i < snake.size(); i++) {
            int[] current = snake.get(i);
            if(current[0] == snakeHead[0] && current[1] == snakeHead[1]) {
                return -1;
            }
        }

        //Eats Food
        if(!foodList.isEmpty()) {
            int[] appearedFood = foodList.get(0);
            if(snakeHead[0] == appearedFood[0] && snakeHead[1] == appearedFood[1]) {
                snake.add(foodList.remove());
                return snake.size() - 1;
            }
        }

        //Normal Move
        int[] newCell = new int[]{snakeHead[0],snakeHead[1]};
        snake.add(newCell);
        snake.remove();
        return snake.size() - 1;
    }

}